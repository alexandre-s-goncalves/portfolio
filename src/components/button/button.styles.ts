import styled from 'styled-components';
import {colors} from 'resources/colors';
import {margin} from 'resources/margins';
import {Text} from 'components/ui/text/Text';
import {Icon} from 'components/icon';

export const ButtonContainer = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  gap: ${margin.xsmall}px;
  padding: ${margin.xsmall}px ${margin.medium}px;
  border-radius: 8px;
  border: none;
  height: 40px;
  transition: all 0.2s ease;
  background-color: ${colors.lightgray};
  cursor: pointer;

  &:hover {
    opacity: 0.9;
  }

  &:active {
    transform: scale(0.98);
  }

  &:disabled {
    cursor: default;

    &:hover {
      opacity: 1;
    }
  }
`;

export const ButtonIcon = styled(Icon).attrs(props => ({
  width: props.width || 24,
  height: props.height || 24,
  color: props.color || colors.white,
}))``;

export const ButtonText = styled(Text).attrs(props => ({
  size: props.size || 14,
  weight: props.weight || 'semi-bold',
  color: props.color || colors.white,
}))`
  white-space: nowrap;
`;
