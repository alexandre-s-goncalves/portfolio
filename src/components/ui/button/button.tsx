import React from 'react';
import {TextProps} from 'components/ui/text/Text.styles';
import * as S from './button.styles';

export interface ButtonProps {
  children: string;
  icon?: React.FunctionComponent<React.SVGProps<SVGSVGElement>>;
  iconPosition?: 'left' | 'right';
  onClick?: () => void;
  disabled?: boolean;
  type?: 'button' | 'submit' | 'reset';
  textProps?: Partial<TextProps>;
  className?: string;
  ariaLabel?: string;
}

export const Button = ({
  children,
  icon,
  iconPosition = 'right',
  onClick,
  disabled = false,
  type = 'button',
  textProps,
  className,
  ariaLabel,
  ...props
}: ButtonProps) => {
  const iconColor = textProps?.color;

  return (
    <S.ButtonContainer
      onClick={onClick}
      disabled={disabled}
      type={type}
      className={className}
      aria-label={ariaLabel || children}
      {...props}>
      {icon && iconPosition === 'left' && <S.ButtonIcon icon={icon} color={iconColor} />}
      <S.ButtonText {...textProps}>{children}</S.ButtonText>
      {icon && iconPosition === 'right' && <S.ButtonIcon icon={icon} color={iconColor} />}
    </S.ButtonContainer>
  );
};
