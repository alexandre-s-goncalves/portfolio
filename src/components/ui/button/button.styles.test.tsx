import React from 'react';
import {render} from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import * as S from './button.styles';
import ArrowSVG from 'assets/icons/right-arrow-icon.svg';

describe('button.styles', () => {
  describe('ButtonContainer', () => {
    test('should match snapshot', () => {
      const {container} = render(<S.ButtonContainer />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should render as button element', () => {
      const {container} = render(<S.ButtonContainer />);
      const element = container.firstChild as HTMLElement;
      expect(element.nodeName).toBe('BUTTON');
    });

    test('should have flex layout', () => {
      const {getByTestId} = render(<S.ButtonContainer data-testid="button" />);
      const button = getByTestId('button');
      expect(button).toHaveStyle('display: flex');
      expect(button).toHaveStyle('align-items: center');
      expect(button).toHaveStyle('justify-content: center');
    });

    test('should have correct height', () => {
      const {getByTestId} = render(<S.ButtonContainer data-testid="button" />);
      const button = getByTestId('button');
      expect(button).toHaveStyle('height: 40px');
    });

    test('should have cursor pointer', () => {
      const {getByTestId} = render(<S.ButtonContainer data-testid="button" />);
      const button = getByTestId('button');
      expect(button).toHaveStyle('cursor: pointer');
    });

    test('should have border radius', () => {
      const {getByTestId} = render(<S.ButtonContainer data-testid="button" />);
      const button = getByTestId('button');
      expect(button).toHaveStyle('border-radius: 8px');
    });

    test('should have correct gap', () => {
      const {getByTestId} = render(<S.ButtonContainer data-testid="button" />);
      const button = getByTestId('button');
      expect(button).toHaveStyle('gap: 8px');
    });

    test('should have lightgray background', () => {
      const {getByTestId} = render(<S.ButtonContainer data-testid="button" />);
      const button = getByTestId('button');
      expect(button).toHaveStyle('background-color: #d3d3d3');
    });

    test('should have default cursor when disabled', () => {
      const {getByTestId} = render(<S.ButtonContainer data-testid="button" disabled />);
      const button = getByTestId('button');
      expect(button).toHaveStyle('cursor: default');
    });
  });

  describe('ButtonIcon', () => {
    test('should match snapshot', () => {
      const {container} = render(<S.ButtonIcon icon={ArrowSVG} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should render SVG element', () => {
      const {container} = render(<S.ButtonIcon icon={ArrowSVG} />);
      const svg = container.querySelector('svg');
      expect(svg).toBeInTheDocument();
    });
  });

  describe('ButtonText', () => {
    test('should match snapshot', () => {
      const {container} = render(<S.ButtonText>Text</S.ButtonText>);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should have default white color', () => {
      const {getByText} = render(<S.ButtonText>Button Text</S.ButtonText>);
      const text = getByText('Button Text');
      expect(text).toHaveStyle('color: #ffffff');
    });

    test('should have correct font size', () => {
      const {getByText} = render(<S.ButtonText>Text</S.ButtonText>);
      const text = getByText('Text');
      expect(text).toHaveStyle('font-size: 14px');
    });

    test('should prevent text wrapping', () => {
      const {getByText} = render(<S.ButtonText>Text</S.ButtonText>);
      const text = getByText('Text');
      expect(text).toHaveStyle('white-space: nowrap');
    });
  });
});
