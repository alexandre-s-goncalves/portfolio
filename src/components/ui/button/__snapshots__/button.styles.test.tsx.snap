// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button.styles ButtonContainer should match snapshot 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 8px;
  padding: 8px 24px;
  border-radius: 8px;
  border: none;
  height: 40px;
  transition: all 0.2s ease;
  background-color: #d3d3d3;
  cursor: pointer;
}

.c0:hover {
  opacity: 0.9;
}

.c0:active {
  transform: scale(0.98);
}

.c0:disabled {
  cursor: default;
}

.c0:disabled:hover {
  opacity: 1;
}

<button
  class="c0"
/>
`;

exports[`button.styles ButtonIcon should match snapshot 1`] = `
<svg
  class=""
  fill="#ffffff"
  height="24"
  style="min-width: 24px; min-height: 24px; max-width: 24px; max-height: 24px;"
  width="24"
/>
`;

exports[`button.styles ButtonText should match snapshot 1`] = `
.c0 {
  font-size: 14px;
  font-family: Inter-Regular;
  color: #ffffff;
  margin-top: 0px;
  font-weight: normal;
  text-transform: none;
  cursor: text;
}

.c1 {
  white-space: nowrap;
  cursor: inherit;
}

<p
  class="c0 c1"
>
  Text
</p>
`;
